test/run/init_app.js

```diff
@@ -0,0 +1,20 @@
+const { spawn } = require( 'child_process' );
+const { BUILT_APP_DIR } = require( './paths' );
+
+function initApp( log ) {
+    const startCmd = './'
+
+    const app = spawn( startCmd + 'WordPress.com', [
+        '--disable-renderer-backgrounding', '--disable-http-cache',
+        '--start-maximized', '--remote-debugging-port=9222' ],
+        { stdio: [ 'ignore', log, log ], detached: true, cwd: BUILT_APP_DIR }
+    );
+
+    app.on( 'error', err => {
+       throw new Error( 'failed to initialize app: ', err );
+    } );
+
+    return app;
+}
```

Comment:
Just playing around here…

```js
function run( cwd, command, args, output ) {
  const app = spawn( command, args, { stdio: [ 'ignore', output, output ], detached: true, cwd } );
  app.on( 'error' => { throw new Error( `failed to initialize ${command}` ); } );

  return app;
}

const app = run( BUILT_APP_DIR, './WordPress.com', [
  '--disable-renderer-backgrounding',
  '--disable-http-cache',
  '--start-maximized',
  '--remote-debugging-port=9222',
], appLog );

await delay(5000);

const driver = run( PROJECT_DIR, 'npx', [
  'chromedriver',
  '--port=9515',
  '--verbose',
], driverLog );
```

There's a part of me feeling like we could make a simple abstraction over running the process and then eliminate these two `init…` files. I'm nervous about having the _straightforward_ logic of starting the tests be obscured by being hidden in a number of split up files.

Even the `appLog` and `driverLog`, for example, could be inlined in `run-e2e-tests.js`

For transparency, my anxiety about this is from the fact that the test runner systems are always one-off projects and have their own language and style. In these cases I definitely value having more stuff in one file where I can see it in my editor all at once over separating the pieces into separate files and modules.

While it may seem to disagree with my play above to introduce the process abstraction I'm still drawing a line between "this is a thing that spawns a process" and "start the app somehow" and "start the driver somehow." Whereas having the startup commands and CLI args in the test runner itself the mechanisms going on are obvious even if there's something hidden in the process abstraction (redirection of `stdout` and `stderr`). The pattern should be obvious enough from the `cmd, [ arg1, arg2 ]` syntax

Reply:
> There's a part of me feeling like we could make a simple abstraction over running the process and then eliminate these two init… files

Agreed - having the various components initialized in separate files creates unnecessary indirection and overhead when navigating the script. I refactored to inline everything, and the proposed refactor for the `driver` and `app` initialization is a great idea. 👍 See refactor [here](https://github.com/Automattic/wp-desktop/pull/778/commits/c1862f5e56cb6e69217fd9f2110621afc1af3fcf#diff-95f02e568cfd6b3a0b1830f8fd2306a1R87).